bip78/src/uri.rs

```diff
@@ -154,4 +158,87 @@ mod tests {
         assert!(Uri::from_str("bitcoin").is_err());
         assert!(Uri::from_str("bitcoin:").is_err());
     }
+
+    #[ignore]
+    #[test]
+    fn test_todo_url_encoded() {
+        let uri = "bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX?amount=1&pj=https://example.com?ciao";
+        assert!(Uri::from_str(uri).is_err(), "pj url should be url encoded");
+    }
+
+    #[ignore]
+    #[test]
+    fn test_todo_valid_url() {
+        let uri = "bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX?amount=1&pj=http://a";
+        assert!(Uri::from_str(uri).is_err(), "pj is not a valid url");
+    }
+
+    #[ignore]
+    #[test]
+    fn test_todo_missing_amount() {
+        let uri = "bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX?pj=https://testnet.demo.btcpayserver.org/BTC/pj";
+        assert!(Uri::from_str(uri).is_ok(), "missing amount should be ok");
+    }
+
+    #[ignore]
+    #[test]
+    fn test_todo_unencrypted() {
+        let uri = "bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX?amount=1&pj=http://example.com";
+        assert!(Uri::from_str(uri).is_err(), "unencrypted connection");
+    }
+
+    #[test]
+    fn test_valid_uris() {
+        let https = "https://example.com";
+        let onion = "http://vjdpwgybvubne5hda6v4c5iaeeevhge6jvo3w2cl6eocbwwvwxp7b7qd.onion";
+
+        let base58 = "bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX";
+        let bech32_upper = "BITCOIN:TB1Q6D3A2W975YNY0ASUVD9A67NER4NKS58FF0Q8G4";
+        let bech32_lower = "bitcoin:tb1q6d3a2w975yny0asuvd9a67ner4nks58ff0q8g4";
+
+        for address in [base58, bech32_upper, bech32_lower].iter() {
+            for pj in [https, onion].iter() {
+                // TODO add with and without amount
+                // TODO shuffle params
+                let uri = format!("{}?amount=1&pj={}", address, pj);
+                assert!(Uri::from_str(&uri).is_ok());
+            }
+        }
+    }
+
+    #[test]
+    fn test_errors() {
+        assert_matches!(
+            Uri::from_str("bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX").unwrap_err(),
+            ParseUriError::PjNotPresent
+        );
+
+        let bitcoinz = "bitcoinz:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX";
+        let bitcoi = "bitcoi:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX";
+        for schema in [bitcoinz, bitcoi] {
+            let uri = Uri::from_str(schema).unwrap_err();
+            let bad_schema = ParseUriError::from(InternalBip21Error::BadSchema(schema.into()));
+            assert_matches!(uri, bad_schema);
+        }
+
+        let parse_err = ParseUriError::from(InternalBip21Error::Address(
+            util::address::Error::Base58(util::base58::Error::BadChecksum(291609738, 694262922)),
```

Comment:
Going into specifics like this seems too much - basically we're testing what the dependency should have tested already. Just knowing it's `InternalBip21Error::Address` should be sufficient.

I don't feel strongly against it, so feel free to keep it but I suggest you don't bother with testing dependency details next time.

Reply:
Yeah, I agree, done. It was so specific because copied from old equality tests but now we have pattern matching